aten/src/ATen/native/SegmentReduce.cpp

```diff
@@ -29,20 +40,41 @@ Tensor _segment_reduce_cpu_kernel(
         const auto* values_data = data.data_ptr<scalar_t>();
         int64_t k = 0;
         for (int64_t i = 0; i < batch_size; ++i) {
-          scalar_t initial_value = initial.has_value()
-              ? initial.value().to<scalar_t>()
-              : std::numeric_limits<scalar_t>::lowest();
+          // ===== step1: initialize starting value
+          scalar_t initial_value;
+          if (initial.has_value()) {
+            initial_value = initial.value().to<scalar_t>();
+          } else if (reduction == SegmentReductionType::MAX) {
+            initial_value = std::numeric_limits<scalar_t>::lowest();
+          } else if (reduction == SegmentReductionType::MEAN) {
+            initial_value = 0;
+          }
+
+          // ===== step2: apply reduction
           for (int64_t j = 0; j < lengths_data[i]; ++j) {
             const auto data = values_data[k];
-            initial_value = at::_isnan(data)
-                ? data
-                : std::max<scalar_t>(initial_value, data);
+            // TODO: There is no need to branch with every element
+            if (reduction == SegmentReductionType::MAX) {
+              initial_value = at::_isnan(data)
+                  ? data
+                  : std::max<scalar_t>(initial_value, data);
+            } else if (reduction == SegmentReductionType::MEAN) {
+              initial_value = at::_isnan(data) ? data : (initial_value + data);
+            }
             k++;
           }
-          // If unsafe is false, check on lengths or indices should cover cases
-          // where lengths for a particular segment is negative. If unsafe
-          // is true, simply set to initial_value for particular reduction
+
+          // ===== step3: finalize reduction
+          TORCH_CHECK(lengths_data[i] >= 0);
+          if (lengths_data[i] == 0 && !initial.has_value()) {
```

Comment:
you don't check for SegmentReductionType here? Would make sense to add a test to catch that

Reply:
if it is 0 length segment and there is no initial provided, we should always set to NAN, regardless of aggregation type, right?

Reply:
I don't know, numpy errors out on max reduction over 0 elements, I guess it's reasonable to either error out or return -infinity for max reduction

Reply:
Discussed offline, will change defaults at the top of the stack.